hedera-node/src/main/java/com/hedera/services/ledger/PureTransferSemanticChecks.java

```diff
@@ -97,33 +99,61 @@ ResponseCodeEnum validateTokenTransferSizes(
 			return OK;
 		}
 
+		return checkTokenTransfersList(tokenTransfersList,
+				areNftsEnabled,
+				maxOwnershipChanges,
+				maxListLen);
+	}
+
+	private ResponseCodeEnum checkTokenTransfersList(final List<TokenTransferList> tokenTransfersList,
+			final boolean areNftsEnabled,
+			final int maxOwnershipChanges,
+			final int maxListLen) {
 		var count = 0;
 		var numOwnershipChanges = 0;
 		for (var scopedTransfers : tokenTransfersList) {
 			final var ownershipChangesHere = scopedTransfers.getNftTransfersCount();
 			if (ownershipChangesHere > 0) {
-				if (!areNftsEnabled) {
-					return NOT_SUPPORTED;
-				}
-				numOwnershipChanges += ownershipChangesHere;
-				if (numOwnershipChanges > maxOwnershipChanges) {
-					return BATCH_SIZE_LIMIT_EXCEEDED;
+				try {
+					numOwnershipChanges = ownershipChanged(areNftsEnabled,
+							numOwnershipChanges,
+							ownershipChangesHere,
+							maxOwnershipChanges);
+				} catch (InvalidTransactionException e) {
+					return e.getResponseCode();
 				}
 			} else {
-				int transferCounts = scopedTransfers.getTransfersCount();
-				if (transferCounts == 0) {
-					return EMPTY_TOKEN_TRANSFER_ACCOUNT_AMOUNTS;
-				}
-				count += transferCounts;
-				if (count > maxListLen) {
-					return TOKEN_TRANSFER_LIST_SIZE_LIMIT_EXCEEDED;
+
+				try {
+					count = checkTokenTransferNum(scopedTransfers, count, maxListLen);
+				} catch (InvalidTransactionException e) {
```

Comment:
Here also imo if we can avoid using exceptions here also, we should. Not only does throwing an exception create an object, but it captures the stack trace, which can be surprisingly expensive. (See [this blog](https://shipilev.net/blog/2014/exceptional-performance/) for a long discussion about the performance impact of exceptions.)

Reply:
Yeah, changed the implementation to avoid throwing exception.